core/src/main/scala/org/apache/spark/sql/delta/commands/VacuumCommand.scala

```diff
@@ -277,6 +288,34 @@ object VacuumCommand extends VacuumCommandImpl with Serializable {
 
 trait VacuumCommandImpl extends DeltaCommand {
 
+  protected def getFilePath(
+      fileAction: FileAction,
+      fs: FileSystem,
+      reservoirBase: Path,
+      relativizeIgnoreError: Boolean): Seq[String] = {
+    val filePath = stringToPath(fileAction.path)
+    val validFileOpt = if (filePath.isAbsolute) {
+      val maybeRelative =
+        DeltaFileOperations.tryRelativizePath(fs, reservoirBase,
+          filePath, relativizeIgnoreError)
+      if (maybeRelative.isAbsolute) {
+        // This file lives outside the directory of the table
+        None
+      } else {
+        Option(pathToString(maybeRelative))
+      }
+    } else {
+      Option(pathToString(filePath))
+    }
+    validFileOpt.toSeq.flatMap { f =>
+      // paths are relative so provide '/' as the basePath.
+      Seq(f).flatMap { file =>
+        val dirs = getAllSubdirs("/", file, fs)
+        dirs ++ Iterator(file)
```

Comment:
Again, I know the original code had this, but this code will produce many duplicate parent paths, one for each file that was deleted from a given directory. Given the extreme latency of delete calls on some cloud storage platforms, it seems like we would want to de-duplicate the file list before making API calls?

Reply:
If we delete only file paths from the state, this problem should not happen right? we can deduplicate the add files also before the major GC, this will not impact the delete API calls but will probably make the major GC faster (should make the left-anti join between discovered files and add files faster but will add a shuffle to deduplicate the data). what do you think?

Reply:
I think it may not add a  shuffle and the exchange will be reused(the de-duplication and the join use the same key). I will double-check it.

Reply:
Either way, the API calls are the slowest part of vacuum, and by a wide margin, so the dedup is worth it even this query does technically slow down a bit.